core/src/G3Map.cxx

```diff
@@ -4,6 +4,201 @@
 #include <container_pybindings.h>
 #include <serialization.h>
 
+/* Special load/save for int64_t. */
+
+static
+int bit_count(std::map<std::string, int64_t> const &d) {
+	// Returns the smallest number N such that all ints in the
+	// vector could be safely expressed as intN_t.  Assumes two's
+	// complement integers.  Return value will be between 1 and
+	// 64.
+	uint64_t mask = 0;
+	for (auto c: d) {
+		if (c.second < 0)
+			mask |= ~c.second;
+		else
+			mask |= c.second;
+	}
+	for (int i=1; i<64; i++) {
+		if (mask == 0)
+			return i;
+		mask >>= 1;
+	}
+	return 64;
+}
+
+template <class A, typename FROM_TYPE, typename TO_TYPE>
+void load_as(A &ar, std::map<std::string, TO_TYPE> &dest) {
+	std::map<std::string, FROM_TYPE> temp;
+	ar & cereal::make_nvp("map", temp);
+	dest.insert(temp.begin(), temp.end());
+}
+
+template <class A, typename FROM_TYPE, typename TO_TYPE>
+void save_as(A &ar, const std::map<std::string, FROM_TYPE> &src) {
+	std::map<std::string, TO_TYPE> temp(src.begin(), src.end());
+	ar & cereal::make_nvp("map", temp);
+}
+
+template <>
+template <class A>
+void G3Map<std::string, int64_t>::load(A &ar, const unsigned v)
+{
+	G3_CHECK_VERSION(v);
+
+	ar & cereal::make_nvp("G3FrameObject",
+			      cereal::base_class<G3FrameObject>(this));
+	int store_bits = 32;
+	if (v >= 2)
+		ar & cereal::make_nvp("store_bits", store_bits);
+
+	switch(store_bits) {
+	case 64:
+		ar & cereal::make_nvp("map",
+				      cereal::base_class<std::map<std::string, int64_t> >(this));
+		break;
+	case 32:
+		load_as<A, int32_t, int64_t>(ar, *this);
+		break;
+	case 16:
+		load_as<A, int16_t, int64_t>(ar, *this);
+		break;
+	case 8:
+		load_as<A, int8_t, int64_t>(ar, *this);
+		break;
+	}
+}
+
+template <>
+template <class A>
+void G3Map<std::string, int64_t>::save(A &ar, const unsigned v) const
+{
+	// v == 2
+	ar & cereal::make_nvp("G3FrameObject",
+			      cereal::base_class<G3FrameObject>(this));
+	// Count the interesting bits, and convert to nearest power of 2.
+	int sig_bits = bit_count(*this);
+	int store_bits = 8;
+	while (store_bits < sig_bits)
+		store_bits *= 2;
+	ar & cereal::make_nvp("store_bits", store_bits);
+	switch(store_bits) {
+	case 8:
+		save_as<A, int64_t, int8_t>(ar, *this);
+		break;
+	case 16:
+		save_as<A, int64_t, int16_t>(ar, *this);
+		break;
+	case 32:
+		save_as<A, int64_t, int32_t>(ar, *this);
+		break;
+	default:
+		ar & cereal::make_nvp("map",
+				      cereal::base_class<std::map<std::string, int64_t> >(this));
+	}
+}
+
+static
+int bit_count_vector(std::map<std::string, std::vector<int64_t> > const &d) {
```

Comment:
Really this is `bit_count_map_vector`, right?

In the case of G3MapVectorInt, do you think it's efficient enough, to compute a single value of store_bits for *all* the data in all the entries?  I would have thought that in many cases you'd have heterogeneous data, from one field to the next, and thus benefit from computing store_bits separately for each entry.

Reply:
Good call, I've rearranged G3MapVectorInt to compress each vector independently.